src/midi/midi_mt32.cpp

```diff
@@ -683,101 +694,124 @@ void MidiHandler_mt32::Close()
+uint16_t MidiHandler_mt32::GetNumPendingFrames()
 {
-	const uint32_t played_frames = total_buffers_played * FRAMES_PER_BUFFER;
-	return service->convertOutputToSynthTimestamp(played_frames +
-	                                              last_played_frame);
+	const auto now_ms = PIC_FullIndex();
+
+	// Wake up the channel and update the last rendered time datum.
+	assert(channel);
+	if (channel->WakeUp()) {
+		last_rendered_ms = now_ms;
+		return 0;
+	}
+	if (last_rendered_ms >= now_ms) {
+		return 0;
+	}
+	// Return the number of frames needed to get current again
+	assert(ms_per_frame > 0.0);
+	const auto elapsed_ms = now_ms - last_rendered_ms;
+	const auto frames     = iround(ceil(elapsed_ms / ms_per_frame));
+	last_rendered_ms += (frames * ms_per_frame);
+	return check_cast<uint16_t>(frames);
 }
 
+// The request to play the channel message is placed in the MIDI work FIFO
 void MidiHandler_mt32::PlayMsg(const uint8_t *msg)
 {
-	assert(channel);
-	channel->WakeUp();
-
-	const auto msg_words = reinterpret_cast<const uint32_t *>(msg);
-	const std::lock_guard<std::mutex> lock(service_mutex);
-	service->playMsgAt(SDL_SwapLE32(*msg_words), GetMidiEventTimestamp());
+	constexpr auto channel_len = 4;
+	std::vector<uint8_t> message(msg, msg + channel_len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Channel};
+	work_fifo.Enqueue(std::move(work));
 }
 
+// The request to play the sysex message is placed in the MIDI work FIFO
 void MidiHandler_mt32::PlaySysex(uint8_t *sysex, size_t len)
 {
-	assert(channel);
-	channel->WakeUp();
-
-	assert(len <= UINT32_MAX);
-	const auto msg_len = static_cast<uint32_t>(len);
-	const std::lock_guard<std::mutex> lock(service_mutex);
-	service->playSysexAt(sysex, msg_len, GetMidiEventTimestamp());
+	std::vector<uint8_t> message(sysex, sysex + len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Sysex};
+	work_fifo.Enqueue(std::move(work));
 }
 
 // The callback operates at the frame-level, steadily adding samples to the
 // mixer until the requested numbers of frames is met.
-void MidiHandler_mt32::MixerCallBack(uint16_t requested_frames)
+void MidiHandler_mt32::MixerCallBack(const uint16_t requested_frames)
 {
-	while (requested_frames) {
-		const auto frames_to_be_played = std::min(GetRemainingFrames(),
-		                                          requested_frames);
-		const auto sample_offset_in_buffer = play_buffer.data() +
-		                                     last_played_frame * 2;
-		channel->AddSamples_sfloat(frames_to_be_played, sample_offset_in_buffer);
-		requested_frames -= frames_to_be_played;
-		last_played_frame += frames_to_be_played;
+	assert(channel);
+
+	constexpr auto warning_percent = 5;
+	if (const auto percent_full = frame_fifo.GetPercentFull(); percent_full < warning_percent) {
+		static auto iteration = 0;
+		if (iteration++ % 100 == 0) {
+			LOG_WARNING("MT32: Audio FIFO is %2d%% full", percent_full);
```

Comment:
So we're issuing a warning if the FIFO is less than 5% full? Was that the intention here, not to log when it's close to full?

Reply:
That's right; the sunny day scenario when the MIDI engine is keeping up and able to keep it's audio frame FIFO full - so when the DOSBox mixer comes knocking for audio frames, they're ready with zero wait - and they're pulled off the top of the FIFO.

It's only when the FIFO gets dangerously low when things get close to become more synchronous - that is, the mixer requests for audio frames actually has to sit and wait some microseconds for the MIDI engine to hand them over (via the FIFO, which will shuffle them immediately).

Reply:
In this case, I think it would be best if the warning message contained the term "buffer underrun" somewhere @kcgen  to make the intent of the message obvious (this is a very common term in audio programming).

Reply:
I agree.

Maybe: `"MT32: Audio buffer underrun (%4.2f%% full)"`?

We can set this as the template to use for the other async audio devices (as we migrate them over).

Reply:
Sounds good to me 👍🏻